designs/memory-management-design.md

```diff
@@ -0,0 +1,275 @@
+    def add(l: Link) -> Link:
+        m: Link = None
+        m = Link()
+        l.next = m
+        return m
+
+x: Link = None
+y: Link = None
+y = x.add()
+```
+**Expected:**
+
+Let `o` be the object referred to by variable `x` <br>
+Let `p` be the object referred to by variable `y`
+```
+assert number of references of o is 1
+assert type of fields in o is [value, pointer]
+assert number of references of p is 2
+assert type of fields in p is [value, pointer]
+```
+
+### 8. Anonymous object deletion
+
+**Case:**
+```
+class Link(object):
+    val: int = 0
+    next: Link = None
+    def add(l: Link) -> Link:
+        l.next = Link()
+        return l.next
+
+x: Link = None
+x.add()
+x = None
+```
+**Expected:**
+```
+assert number of references of any object is 0
+```
+
+### 9. Simple linked cycle
+**Case:**
+```
+class Link(object):
+    val: int = 0
+    next: Link = None
+
+x: Link = None
+y: Link = None
+x = Link()
+y = Link()
+x.next = y
+y.next = x
+```
+**Expected:**
+
+Let `o` be the object referred to by variable `x` <br>
+Let `p` be the object referred to by variable `y`
+```
+assert number of references of o is 1
+assert type of fields in o is [value, pointer]
+assert number of references of p is 1
+assert type of fields in p is [value, pointer]
+```
+
+### 10. Simple deletion in cycle
+**Case:**
+```
+class Link(object):
+    val: int = 0
+    next: Link = None
+
+x: Link = None
+y: Link = None
+x = Link()
+y = Link()
+x.next = y
+y.next = x
+
+x = None
+```
+**Expected:**
+
+Let `o` be the object referred to by variable `x` <br>
+Let `p` be the object referred to by variable `y`
+```
+assert number of references of o is 0
+assert type of fields in o is [value, pointer]
+assert number of references of p is 1
+assert type of fields in p is [value, pointer]
+```
+
+## Changes to IR
+
+- ### Addition of `ref` value
+```
+export type Value<A> =
+...
+| { a?: A, tag: "ref"}
+```
+Currently there is no way of differentiating between pointers and values stored in an object. As seen in test case *8: Anonymous object deletion*, we need to follow references stored in an object recursively so that we can update their respective reference counts. This is not possible without knowing which field represents a reference/pointer.
+
+- ### Changes to `alloc` expression
+```
+export type Expr<A> =
+  ...
+  | {  a?: A, tag: "alloc", amount: Value<A>[] }
+```
+ To follow references and to deallocate memory on the heap, we need the size and the type information of an object which is stored on it. This expression, allows the memory manager to understand how many bytes are to be allocated/ deleted when the object is garbage collected and what will be the type of data in a field.
+
+- ### Remove BigInt type
+We feel that the implementation of `bigint` type would probably be as a contiguous array of `int` and could just be made into an `object`. This relates to our implementation since we would not have to account for a special case.
```

Comment:
Ah, interesting. So you're proposing to not have `bigint` at the IR level and just make it the group's problem to figure out how to alloc it. I buy that.